libass/ass_directwrite.c

```diff
@@ -668,50 +670,49 @@ static void add_font(IDWriteFont *font, IDWriteFontFamily *fontFamily,
 }
 
 /*
- * Scan every system font on the current machine and add it
- * to the libass lookup. Stores the FontPrivate as private data
- * for later memory reading
+ * When a new font name is requested, called to load that font from Windows
  */
-static void scan_fonts(IDWriteFactory *factory,
-                       ASS_FontProvider *provider)
+static void match_fonts(void *priv, ASS_Library *lib,
+                        ASS_FontProvider *provider, char *name)
 {
-    HRESULT hr = S_OK;
-    IDWriteFontCollection *fontCollection = NULL;
+    ProviderPrivate *provider_priv = (ProviderPrivate *)priv;
     IDWriteFont *font = NULL;
-    hr = IDWriteFactory_GetSystemFontCollection(factory, &fontCollection, FALSE);
+    IDWriteFontFamily *fontFamily = NULL;
+    LOGFONTW lf;
+    HRESULT hr = S_OK;
 
-    if (FAILED(hr) || !fontCollection)
+    memset(&lf, 0, sizeof(LOGFONTW));
+    lf.lfEscapement = GM_COMPATIBLE;
+    lf.lfWeight = FW_DONTCARE;
+    lf.lfCharSet = DEFAULT_CHARSET;
+    lf.lfOutPrecision = OUT_TT_PRECIS;
+    lf.lfClipPrecision = CLIP_DEFAULT_PRECIS;
+    lf.lfQuality = DEFAULT_QUALITY;
+    lf.lfPitchAndFamily = DEFAULT_PITCH;
+
+    // lfFaceName can hold up to LF_FACESIZE wchars; truncate longer names
+    MultiByteToWideChar(CP_UTF8, 0, name, -1, lf.lfFaceName, LF_FACESIZE-1);
+    lf.lfFaceName[LF_FACESIZE-1] = 0;
```

Comment:
And I’ve just realized this line is redundant because the struct is prefilled with zeros.

Reply:
This line is to make sure this field is null-terminated. `wcsncpy` will not append a null character if source string is longer than dest buffer. Since we get rid of that intermediate buffer which is not prefilled with zeros, we can just remove this line.

Reply:
Ah, right, I’d forgotten this was a remnant of the intermediate buffer which wasn’t zero-filled.